src/omp/backend.f90

```diff
@@ -66,57 +67,64 @@ function init(globs, allocator) result(backend)
-    allocate (backend%du_recv_e(SZ, 1, n_block))
-    allocate (backend%dud_send_s(SZ, 1, n_block))
-    allocate (backend%dud_send_e(SZ, 1, n_block))
-    allocate (backend%dud_recv_s(SZ, 1, n_block))
-    allocate (backend%dud_recv_e(SZ, 1, n_block))
-    allocate (backend%d2u_send_s(SZ, 1, n_block))
-    allocate (backend%d2u_send_e(SZ, 1, n_block))
-    allocate (backend%d2u_recv_s(SZ, 1, n_block))
-    allocate (backend%d2u_recv_e(SZ, 1, n_block))
+    backend%mesh => mesh
+    n_groups = maxval([ backend%mesh%get_n_groups(DIR_X), &
+                        backend%mesh%get_n_groups(DIR_Y), & 
+                        backend%mesh%get_n_groups(DIR_Z) ])
+
+    allocate (backend%u_send_s(SZ, n_halo, n_groups))
+    allocate (backend%u_send_e(SZ, n_halo, n_groups))
+    allocate (backend%u_recv_s(SZ, n_halo, n_groups))
+    allocate (backend%u_recv_e(SZ, n_halo, n_groups))
+    allocate (backend%v_send_s(SZ, n_halo, n_groups))
+    allocate (backend%v_send_e(SZ, n_halo, n_groups))
+    allocate (backend%v_recv_s(SZ, n_halo, n_groups))
+    allocate (backend%v_recv_e(SZ, n_halo, n_groups))
+    allocate (backend%w_send_s(SZ, n_halo, n_groups))
+    allocate (backend%w_send_e(SZ, n_halo, n_groups))
+    allocate (backend%w_recv_s(SZ, n_halo, n_groups))
+    allocate (backend%w_recv_e(SZ, n_halo, n_groups))
+
+    allocate (backend%du_send_s(SZ, 1, n_groups))
+    allocate (backend%du_send_e(SZ, 1, n_groups))
+    allocate (backend%du_recv_s(SZ, 1, n_groups))
+    allocate (backend%du_recv_e(SZ, 1, n_groups))
+    allocate (backend%dud_send_s(SZ, 1, n_groups))
+    allocate (backend%dud_send_e(SZ, 1, n_groups))
+    allocate (backend%dud_recv_s(SZ, 1, n_groups))
+    allocate (backend%dud_recv_e(SZ, 1, n_groups))
+    allocate (backend%d2u_send_s(SZ, 1, n_groups))
+    allocate (backend%d2u_send_e(SZ, 1, n_groups))
+    allocate (backend%d2u_recv_s(SZ, 1, n_groups))
+    allocate (backend%d2u_recv_e(SZ, 1, n_groups))
 
   end function init
 
   subroutine alloc_omp_tdsops( &
-    self, tdsops, n, dx, operation, scheme, &
+    self, tdsops, dir, operation, scheme, &
     n_halo, from_to, bc_start, bc_end, sym, c_nu, nu0_nu &
     )
     implicit none
 
     class(omp_backend_t) :: self
     class(tdsops_t), allocatable, intent(inout) :: tdsops
-    integer, intent(in) :: n
-    real(dp), intent(in) :: dx
+    integer, intent(in) :: dir
     character(*), intent(in) :: operation, scheme
     integer, optional, intent(in) :: n_halo
     character(*), optional, intent(in) :: from_to, bc_start, bc_end
     logical, optional, intent(in) :: sym
     real(dp), optional, intent(in) :: c_nu, nu0_nu
+    integer :: data_loc, tds_n
+    real(dp) :: delta
 
     allocate (tdsops_t :: tdsops)
 
     select type (tdsops)
     type is (tdsops_t)
-      tdsops = tdsops_t(n, dx, operation, scheme, n_halo, from_to, &
+      data_loc = get_operation_data_loc(operation, scheme, from_to)
+      tds_n = self%mesh%get_n(dir, data_loc)
```

Comment:
If we get the dimensions in 3D mesh both for `vert` and `cell` first, then we can get the number of cells or vertices along a `dir` easily.
```
n_cells(:) = mesh%get_mesh_dims(CELL)
n_verts(:) = mesh%get_mesh_dims(VERT)
if (from_to == 'v2p') then
  tds_n = n_cells(dir)
else
  tds_n = n_verts(dir)
end if
```

or we can also use `mesh%get_n_dir`
```
if (from_to == 'v2p') then
  tds_n = mesh%get_n_dir(dir, CELL)
else
  tds_n = mesh%get_n_dir(dir, VERT)
end if
```

Reply:
But `from_to` is optional, when it isn't defined is it fine to have `tds_n = mesh%get_n_dir(dir, VERT)`?

Reply:
Yes, if not defined we assume its `v2v`, so we need `VERT` dimensions in that case too.

Reply:
ok, that makes things a lot easier than I thought